projecttacoma/synthea
# Added Measurement Period Guard to EXM Modules

# Summary
The EXM111, EXM125, EXM104, EXM105, EXM130, and EXM506 JSON files were updated with measurement period guards to enforce the measurement period in Synthea.

## New Behavior
The EXM125 JSON module previously included code to create a measurement period guard in Synthea. This measurement period guard does not format well in the Synthea Module Builder, but its attribute logic is correct. Using that code as guidance, measurement period guards were added to EXM111, EXM104, EXM105, EXM130, and EXM506. 

The measurement period guard logic in EXM125 accounts for the measurement period start but not the measurement period end. Because the events in Synthea typically occur immediately, the measurement period end was not added to the guard logic at this time, as it will not alter the guard's functionality. However, EXM105 and EXM111 both call for inpatient encounters with a length of stay less than or equal to 120 days that ends during the measurement period. This means the encounter must start on or before September 2. In a previous PR, we checked whether the month was < 10 in a separate state, which is not as accurate as it could be. Therefore, that logic was included in the measurement period guard to account for the length of stay. EXM104 and EXM506 also have similar requirements, so the September 2 encounter date was included in the measurement period guard logic, although these modules currently use multiple-year delays that counteract the guard. In the future, we may need to remove these delays to make these modules more accurate to their ECQI descriptions.

For some of the modules, the age guard had appeared after the measurement period guard. If this was the case, the two guards were swapped because the patient should be of the appropriate age at the start of the measurement period.

## Code Changes
For all the modules, a guard state called `Measurement_Period_Guard` was added to check that the date is >= the start of the measurement period (1/1/2019), which is captured in the `Start of Measurement Period` attribute. For modules without a measurement period end attribute, a SetAttribute state called `Set_Measurement_Period_End` was added. Note that this is for future use and is not used in the `Measurement_Period_Guard` logic. 
Although the measurement period end was not incorporated into the measurement period guard, its attribute was added to all the modules that were missing it, just so that we can use it in the future, if needed. 

For modules where the length of stay needs to occur before the measurement period end, an attribute called `End of Encounter Period` was added, which captures the September 2, 2019 cutoff. In the `Measurement_Period_Guard` logic, the guard checks that the date is <= the end of the encounter period.

# Testing Guidance

- First, enter each module's JSON into the Synthea Module Builder to see if the order of the states make sense. The general setup has the attribute states at the top, followed by guards, including the `Measurement_Period_Guard`. Because `Measurement_Period_Guard` uses attributes in its logic (which is not fully supported by the Module Builder), you should see something like `Allow if Symptom: 'undefined' >= undefined'.
- Go to `synthea/src/main/resources` and open `synthea.properties`. Go to line 64 and edit `generate.log_patients.detail = simple` to `generate.log_patients.detail = detailed`. This will allow for detailed patient information to appear in the terminal, including a list of all the states that the patient reaches in the module. 
- Go to `State.java` (under `synthea/src/main/java/org/mitre/synthea/engine`). Add `import org.mitre.synthea.export.ExportHelper;` to the top of the file. Then, update the `toString()` method so that it looks like this:
`[...]`
  `public String toString() {`
    `return this.getClass().getSimpleName() + " '" + name + "' " + ExportHelper.iso8601Timestamp(this.entered) + " --> " + (this.exited == null ? "" : ExportHelper.iso8601Timestamp(this.exited)) ;
  }`
This addition allows you to see the timestamps at each state when you are viewing the detailed output. You will be able to see when the patient enters the state, followed by the --> arrow, followed by the timestamp at which the patient leaves the state. (Shoutout to Dylan Hall for helping with this!)
- Run Synthea with all six modules with the following commands: `./run_synthea -p 1 -m "stroke_exm_104_r4"`, `./run_synthea -p 1 -m "stroke_exm_105_r4"`, `./run_synthea -p 1 -m "EXM506*"`, `./run_synthea -p 1 -m "EXM130*-R4"`, `./run_synthea -p 1 -m "EXM125*-R4"`, `./run_synthea -p 1 -m "EXM111*"`. You should receive detailed output for the patient. Scroll under `ATTRIBUTES` to the list containing all the states that the patient reaches, which is in reverse chronological order. If the patient passes through the `Measurement_Period_Guard`, check that the exit date is within the 2019 calendar year. For EXM104 and EXM506, you will see that some of the states' dates are later than 2019 if they went through a delay state. This is expected.
